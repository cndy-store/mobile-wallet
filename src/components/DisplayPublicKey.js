import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { chunk } from 'lodash';
import { View, Button, StyleSheet, Text, Platform } from 'react-native';

const pairsPerRow = 4;

const needsBold = index => index % 2 === 0;
const needsNewLine = index => index === pairsPerRow - 1;

const fontFamily = Platform.OS === 'ios' ? 'Courier' : 'monospace';

class DisplayPublicKey extends Component {
  pairsToText(pairs) {
    return pairs.map((pair, index) => {
      const key = `${pair}-${index}`;
      const style = needsBold(index) ? styles.bold : styles.normal;
      const newLine = needsNewLine(index) ? '\n' : '';

      return (
        <Text key={key} style={style}>
          {pair}
          {newLine}
        </Text>
      );
    });
  }

  keyParts() {
    const chars = this.props.publicKey.split('');
    const pairs = chunk(chars, 2).map(pair => pair.join(''));
    const perRow = chunk(pairs, 4);

    return perRow.map((pairs, index) => {
      return this.pairsToText(pairs);
    });
  }

  render() {
    return (
      <View style={{ alignSelf: 'center' }}>
        <Text numberOfLines={7} style={styles.publicKey}>
          {this.keyParts()}
        </Text>
      </View>
    );
  }
}

DisplayPublicKey.propTypes = {
  publicKey: PropTypes.string.isRequired
};

const styles = StyleSheet.create({
  publicKey: {
    fontSize: 24
  },
  normal: {
    fontFamily,
    letterSpacing: 5,
    fontSize: 30
  },
  bold: {
    fontFamily,
    letterSpacing: 5,
    fontSize: 30,
    fontWeight: 'bold'
  }
});

export default DisplayPublicKey;
