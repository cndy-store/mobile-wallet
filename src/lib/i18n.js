import NativeLocale from 'react-native-locale';

const localeSettings = NativeLocale.constants();

export { localeSettings };
