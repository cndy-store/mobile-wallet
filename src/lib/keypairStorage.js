import * as Keychain from 'react-native-keychain';

const getItem = async _key => {
  const credentials = await Keychain.getGenericPassword();
  return Promise.resolve(credentials.password);
};

const setItem = (key, value) => {
  return Keychain.setGenericPassword(key, value);
};

const removeItem = _key => {
  return Keychain.resetGenericPassword();
};

export { getItem, setItem, removeItem };
