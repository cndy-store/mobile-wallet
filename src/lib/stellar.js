import axios from 'axios';
import StellarSdk from '../vendor/stellar-base-0.7.7';
StellarSdk.Network.useTestNetwork();

const assetCode = 'CNDY';
const issuerPublicKey =
  'GCJKC2MI63KSQ6MLE6GBSXPDKTDAK43WR522ZYR3F34NPM7Z5UEPIZNX';

const asset = new StellarSdk.Asset(assetCode, issuerPublicKey);

const axiosInstance = axios.create({
  baseURL: 'https://horizon-testnet.stellar.org',
  timeout: 10 * 1000
});

module.exports = {
  axios: axiosInstance,
  asset,
  StellarSdk
};
