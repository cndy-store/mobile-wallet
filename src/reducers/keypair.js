import {
  KEYPAIR_STARTS_PROCESSING,
  KEYPAIR_ERROR,
  KEYPAIR_LOAD,
  KEYPAIR_SAVE,
  KEYPAIR_REMOVE
} from '../actions/keypair';

const defaultState = {
  isProcessing: false,
  error: null,
  keypair: null
};

export default function keypair(state = defaultState, action) {
  switch (action.type) {
    case KEYPAIR_STARTS_PROCESSING:
      return { ...state, isProcessing: action.isProcessing };
    case KEYPAIR_ERROR:
      return {
        ...state,
        error: action.error,
        isProcessing: action.isProcessing
      };
    case KEYPAIR_LOAD:
    case KEYPAIR_SAVE:
    case KEYPAIR_REMOVE:
      return {
        ...state,
        keypair: action.keypair,
        error: action.error,
        isProcessing: action.isProcessing
      };
    default:
      return state;
  }
}
