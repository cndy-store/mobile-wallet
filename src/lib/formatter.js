import { padEnd, reduceRight } from 'lodash';
import { localeSettings } from './i18n';

const delimiter = localeSettings.decimalSeparator;
const finalDelimiter = '.';
const maxPrecisionLength = 7;

const allowedChars = new RegExp(`^[0-9${delimiter}${finalDelimiter}]+`);
const delimiterSplit = new RegExp(`[${delimiter}${finalDelimiter}]+`);

const sanitizeOngoingAmountInput = input => {
  const onlyAllowedChars = input.match(allowedChars);
  if (!onlyAllowedChars) return '';
  if (onlyAllowedChars[0] !== input) return onlyAllowedChars[0];

  const [precision, ...rest] = input.split(delimiterSplit);
  if (!precision.length) return '';
  if (!rest.length) return precision;

  let scale = rest.join('');
  if (scale.length > maxPrecisionLength) {
    scale = scale.substr(0, maxPrecisionLength);
  }

  return [precision, scale].join(delimiter);
};

const ensureScale = input => {
  const [precision, scale] = input.split(delimiter);
  const paddedScale = padEnd(scale || '', 2, '0');

  return `${precision}${delimiter}${paddedScale}`;
};

const parseTransactionAmount = input => {
  const sanitized = sanitizeOngoingAmountInput(input);
  if (sanitized === '') return null;
  const withScale = ensureScale(sanitized);

  const withCorrectDelimiter = withScale.replace(delimiter, finalDelimiter);
  return withCorrectDelimiter;
};

const optimizeScale = scale => {
  const scaleNumbers = scale.split('');
  const neededNumbers = reduceRight(
    scaleNumbers,
    (acc, number) => {
      if (acc.length || number !== '0') {
        acc.push(number);
      }
      return acc;
    },
    []
  );

  return neededNumbers.reverse().join('');
};

// this method is to be used with normalized amount strings
const shortFormat = input => {
  const [precision, scale] = input.split(finalDelimiter);
  const optimizedScale = optimizeScale(scale);
  const paddedScale = padEnd(optimizedScale, 2, '0');

  return `${precision}${delimiter}${paddedScale}`;
};

export { sanitizeOngoingAmountInput, parseTransactionAmount, shortFormat };
