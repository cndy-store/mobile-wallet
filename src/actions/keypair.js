import { getItem, setItem, removeItem } from '../lib/keypairStorage';
import { StellarSdk } from '../lib/stellar';
const { Keypair, StrKey } = StellarSdk;

const STORAGE_KEY = 'STELLAR_SECRET';

export const KEYPAIR_STARTS_PROCESSING = 'KEYPAIR_STARTS_PROCESSING';
export const KEYPAIR_ERROR = 'KEYPAIR_ERROR';
export const KEYPAIR_LOAD = 'KEYPAIR_LOAD';
export const KEYPAIR_SAVE = 'KEYPAIR_SAVE';
export const KEYPAIR_REMOVE = 'KEYPAIR_REMOVE';

const chomp = str => str.replace(/\n|\r| /g, '');
const isValidSecret = secret => StrKey.isValidEd25519SecretSeed(secret);

export const keypairProcessing = bool => ({
  type: KEYPAIR_STARTS_PROCESSING,
  isProcessing: bool
});

export const keypairErrored = ({ error }) => ({
  type: KEYPAIR_ERROR,
  error
});

export const keypairLoadSucceded = ({ secret }) => {
  const keypair = secret ? Keypair.fromSecret(secret) : null;

  return {
    type: KEYPAIR_LOAD,
    keypair: keypair,
    error: null
  };
};

export const keypairSaveSucceded = ({ secret }) => {
  const keypair = Keypair.fromSecret(secret);

  return {
    type: KEYPAIR_SAVE,
    keypair: keypair,
    error: null
  };
};

export const keypairRemoveSucceded = () => ({
  type: KEYPAIR_REMOVE,
  keypair: null,
  error: null
});

export const loadKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));
    try {
      const secret = await getItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairLoadSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const saveKeypair = secret => {
  secret = chomp(secret);

  return async dispatch => {
    dispatch(keypairProcessing(true));

    if (!isValidSecret(secret)) {
      const error = new Error('Invalid Secret');
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }

    try {
      await setItem(STORAGE_KEY, secret);
      dispatch(keypairProcessing(false));
      dispatch(keypairSaveSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const removeKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));

    try {
      await removeItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairRemoveSucceded());
      return Promise.resolve();
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};
