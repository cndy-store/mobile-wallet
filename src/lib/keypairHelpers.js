import { asset, StellarSdk } from '../lib/stellar';
import bip21 from 'bip21';

const { StrKey } = StellarSdk;

const scheme = asset.getCode().toLowerCase();
const secretScheme = `${scheme}secret`;

const isValidPublicKey = publicKey => StrKey.isValidEd25519PublicKey(publicKey);
const isValidSecret = secret => StrKey.isValidEd25519SecretSeed(secret);

const encodePublicKey = (publicKey, options = {}) => {
  if (!isValidPublicKey(publicKey)) return null;
  options.scheme = scheme;
  return bip21.encode(publicKey, options);
};

const decodePublicKey = url => {
  let decoded;
  try {
    decoded = bip21.decode(url, scheme);
  } catch (e) {
    return null;
  }

  const publicKey = decoded.address;
  if (!isValidPublicKey(publicKey)) return null;

  return {
    publicKey,
    options: decoded.options
  };
};

const encodeSecret = (secret, options = {}) => {
  if (!isValidSecret(secret)) return null;
  options.scheme = secretScheme;
  return bip21.encode(secret, options);
};

const decodeSecret = url => {
  let decoded;
  try {
    decoded = bip21.decode(url, secretScheme);
  } catch (e) {
    return null;
  }

  const secret = decoded.address;
  if (!isValidSecret(secret)) return null;

  return {
    secret,
    options: decoded.options
  };
};

export {
  isValidPublicKey,
  isValidSecret,
  encodePublicKey,
  decodePublicKey,
  encodeSecret,
  decodeSecret
};
