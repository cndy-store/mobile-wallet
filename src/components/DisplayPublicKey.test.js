import React from 'react';
import { render } from '../__tests__/renderer';
import { asset } from '../lib/stellar';
import DisplayPublicKey from './DisplayPublicKey';

import { publicKey } from '../__tests__/fixtures/keypair';

it('renders a public key correctly', () => {
  const { toJSON } = render(<DisplayPublicKey publicKey={publicKey} />);
  expect(toJSON()).toMatchSnapshot();
});
