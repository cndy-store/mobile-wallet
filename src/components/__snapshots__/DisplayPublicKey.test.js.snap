// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a public key correctly 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    numberOfLines={7}
    style={
      Object {
        "fontSize": 24,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      GC
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      HS
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      UH
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      OK
      

    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      NZ
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      QA
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      KT
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      UL
      

    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      V5
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      YT
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      MG
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      5C
      

    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      KU
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      2N
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      KE
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      7U
      

    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      UQ
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      VN
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      6J
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      T2
      

    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      JY
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      73
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      37
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      CR
      

    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      BI
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      AZ
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "fontWeight": "bold",
          "letterSpacing": 5,
        }
      }
    >
      6L
      
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontFamily": "Courier",
          "fontSize": 30,
          "letterSpacing": 5,
        }
      }
    >
      XK
      

    </Text>
  </Text>
</View>
`;
